File: lib/config/auction_analytics.dart
/// Firestore: `analytics_events/{autoId}`
///
/// **Schema**
/// | Field | Type | Notes |
/// |-------|------|--------|
/// | `eventType` | string | `bid_placed` \| `auction_viewed` \| `user_outbid` \| `auction_won` |
/// | `userId` | string | Actor (bidder, viewer, outbid user, or winner) |
/// | `lotId` | string | `lots/{lotId}` document id |
/// | `timestamp` | timestamp | Server time (`FieldValue.serverTimestamp()` or Admin SDK) |
///
/// **Writers**
/// - `auction_viewed`: Flutter (signed-in users), rules-restricted.
/// - `bid_placed`, `user_outbid`, `auction_won`: Cloud Functions (Admin SDK).
abstract final class AuctionAnalytics {
  AuctionAnalytics._();

  static const String collection = 'analytics_events';

  static const String bidPlaced = 'bid_placed';
  static const String auctionViewed = 'auction_viewed';
  static const String userOutbid = 'user_outbid';
  static const String auctionWon = 'auction_won';
}

File: functions/src/auctionAnalytics.ts
/**
 * Auction analytics append-only events → `analytics_events`.
 * All eventType values are lowercase snake_case.
 */
import type { Firestore } from "firebase-admin/firestore";
import { FieldValue } from "firebase-admin/firestore";

export const ANALYTICS_EVENTS_COLLECTION = "analytics_events";

export const AuctionAnalyticsEventType = {
  BID_PLACED: "bid_placed",
  AUCTION_VIEWED: "auction_viewed",
  USER_OUTBID: "user_outbid",
  AUCTION_WON: "auction_won",
} as const;

export type AuctionAnalyticsEventTypeName =
  (typeof AuctionAnalyticsEventType)[keyof typeof AuctionAnalyticsEventType];

/**
 * Firestore document shape (fixed fields only):
 * - eventType: string
 * - userId: string
 * - lotId: string
 * - timestamp: server timestamp
 */
export async function recordAuctionAnalyticsEvent(
  db: Firestore,
  args: {
    eventType: AuctionAnalyticsEventTypeName | string;
    userId: string;
    lotId: string;
  }
): Promise<void> {
  const userId = args.userId.trim();
  const lotId = args.lotId.trim();
  if (!userId || !lotId) return;

  try {
    await db.collection(ANALYTICS_EVENTS_COLLECTION).add({
      eventType: args.eventType,
      userId,
      lotId,
      timestamp: FieldValue.serverTimestamp(),
    });
  } catch (e) {
    console.error("[auctionAnalytics] write failed", args.eventType, e);
  }
}
